client/src/main/java/io/pravega/client/netty/impl/ConnectionPoolImpl.java

```diff
@@ -115,19 +114,20 @@ public ConnectionPoolImpl(ClientConfig clientConfig) {
 
         final Connection connection;
         if (suggestedConnection.isPresent() && (prunedConnectionList.size() >= clientConfig.getMaxConnectionsPerSegmentStore() || isUnused(suggestedConnection.get()))) {
-            log.debug("Reusing connection: {}", suggestedConnection.get());
-            Connection oldConnection = suggestedConnection.get();
-            prunedConnectionList.remove(oldConnection);
-            connection = new Connection(oldConnection.getUri(), oldConnection.getFlowHandler());
+            log.info("Reusing connection: {}", suggestedConnection.get());
+            connection = suggestedConnection.get();
         } else {
             // create a new connection.
-            log.debug("Creating a new connection to {}", location);
-            CompletableFuture<FlowHandler> flowHandlerFuture = establishConnection(location);
-            connection = new Connection(location, flowHandlerFuture);
+            log.info("Creating a new connection to {}", location);
+            final AppendBatchSizeTracker batchSizeTracker = new AppendBatchSizeTrackerImpl();
+            final FlowHandler handler = new FlowHandler(location.getEndpoint(), batchSizeTracker);
+            CompletableFuture<Void> establishedFuture = establishConnection(location, handler);
+            connection = new Connection(location, handler, establishedFuture);
+            prunedConnectionList.add(connection);
         }
-        prunedConnectionList.add(connection);
+        ClientConnection result = connection.getFlowHandler().createFlow(flow, rp);
         connectionMap.put(location, prunedConnectionList);
-        return connection.getFlowHandler().thenApply(flowHandler -> flowHandler.createFlow(flow, rp));
+        return connection.getConnected().thenApply(v -> result);
```

Comment:
We should create a `Flow` only once the connection is established.
i.e. 
```
 connection.getConnected().thenApply(v -> connection.getFlowHandler().createFlow(flow, rp));
```
Is there any reason it is being done before?

Reply:
That is the race condition causing the tests to fail that this PR is attempting to fix. See the attached issue.